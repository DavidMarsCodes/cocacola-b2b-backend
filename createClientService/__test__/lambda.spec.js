const handler = require('../lambda');
const { clientMock } = require('./__mocks__/clients');

describe('Lambda Function', () => {
  it('You should save an order in the dynamoDB database.', async () => {

    const event = {
      transactionId: '111111',
      organizationId: '3043',
      cpgId: '001',
      countryId: 'AR',
      clients: [
        {
          clientId: '001AR3043-ERP-TEST-A',
          fiscalId: '227777777722',
          fiscalName: 'Test SA',
          fantasyName: 'TEST-A SA',
          parentClientId: '8888888',
          erpClientId: 'ERP_TEST-A',
          salesOrganization: 'Test',
          distributionChannel: 'Masivo',
          address: {
            street: 'Salguero',
            doorNumber: '227',
            state: 'Buenos Aires',
            city: 'San Isidro',
            zipCode: '1609',
            observations: 'El unico lugar que vale la pena',
          },
          locks: [
            { erpLockId: '001AR3043-LOCK-DELETED' },
          ],
        },
      ],
      createdTime: expect.any(String),
    };

    const awsRepositories = { Repository: { create: jest.fn().mockResolvedValue({ ok: true }) } };

    const getTableName = jest.fn().mockResolvedValue({ tableName: 'devSuggestedProductToUpsert' });

    const experts = { client: { validateUpsertClient: jest.fn() } };

    const res = {
      sendStatus: jest.fn().mockReturnValue({ httpStatus: 201 }),
      error: jest.fn().mockResolvedValue({
        httpStatus: 404,
        ok: true,
        code: 1001,
        errorType: 'Not Found',
        message: 'descripcion del error',
      }),
    };

    const { createClient } = handler({ awsRepositories, getTableName, experts, res });
    await createClient(event);

    expect(experts.client.validateUpsertClient.mock.calls).toMatchSnapshot([
      [
        event,
      ],
    ]);

    expect(awsRepositories.Repository.create.mock.calls).toEqual([
      [
        'devSuggestedProductToUpsert',
        event,
      ],
    ]);

    expect(res.sendStatus.mock.calls).toEqual([
      [
        201,
      ],
    ]);
  });

  it('It should answer a validation error in case any of the required parameters are invalid.', async () => {

    const event = {
      transactionId: '111111',
      organizationId: '3043',
      cpgId: '001',
      countryId: 'AR',
      clients: [
        {
          clientId: '001AR3043-ERP-TEST-A',
          fiscalId: '227777777722',
          fiscalName: 'Test SA',
          fantasyName: 'TEST-A SA',
          parentClientId: '8888888',
          erpClientId: 'ERP_TEST-A',
          salesOrganization: 'Test',
          distributionChannel: 'Masivo',
          address: {
            street: 'Salguero',
            doorNumber: '227',
            state: 'Buenos Aires',
            city: 'San Isidro',
            zipCode: '1609',
            observations: 'El unico lugar que vale la pena',
          },
          locks: [
            { erpLockId: '001AR3043-LOCK-DELETED' },
          ],
        },
      ],
      createdTime: expect.any(String),
    };

    const awsRepositories = { Repository: { create: jest.fn().mockResolvedValue({ ok: true }) } };

    const getTableName = jest.fn().mockResolvedValue({ tableName: 'devSuggestedProductToUpsert' });

    const experts = {
      client: {
        validateUpsertClient: jest.fn(data => {
          throw {
            customError: true,
            getData: jest.fn().mockReturnValue({
              httpStatus: 400,
              status: 'error',
              code: 100,
              msg: 'Validation_server_error',
              type: 'Validation_error',
            }),
          };
        }),
      },
    };

    const res = {
      sendStatus: jest.fn().mockReturnValue({ httpStatus: 201 }),
      error: jest.fn((message = '', code = 0, errorType = 'Internal_Server_Error', status = 500) => ({
        httpStatus: status,
        ok: false,
        code,
        errorType,
        message,
      })),
    };

    const { createClient } = handler({ awsRepositories, getTableName, experts, res });
    await createClient(event);

    expect(getTableName.mock.calls).toEqual([ ]);

    expect(experts.client.validateUpsertClient.mock.calls).toMatchSnapshot([
      [
        event,
      ],
    ]);

    expect(awsRepositories.Repository.create.mock.calls).toEqual([]);

    expect(res.sendStatus.mock.calls).toEqual([]);

    expect(res.error.mock.calls).toEqual([
      [
        'Validation_server_error',
        100,
        'Validation_error',
        400,

      ],
    ]);

  });

  it('It should return defualt error', async () => {

    const event = {
      transactionId: '111111',
      organizationId: '3043',
      cpgId: '001',
      countryId: 'AR',
      clients: [
        {
          clientId: '001AR3043-ERP-TEST-A',
          fiscalId: '227777777722',
          fiscalName: 'Test SA',
          fantasyName: 'TEST-A SA',
          parentClientId: '8888888',
          erpClientId: 'ERP_TEST-A',
          salesOrganization: 'Test',
          distributionChannel: 'Masivo',
          address: {
            street: 'Salguero',
            doorNumber: '227',
            state: 'Buenos Aires',
            city: 'San Isidro',
            zipCode: '1609',
            observations: 'El unico lugar que vale la pena',
          },
          locks: [
            { erpLockId: '001AR3043-LOCK-DELETED' },
          ],
        },
      ],
      createdTime: expect.any(String),
    };

    const awsRepositories = { Repository: { create: jest.fn().mockRejectedValue({ ok: false }) } };

    const getTableName = jest.fn().mockResolvedValue({ tableName: 'devSuggestedProductToUpsert' });

    const experts = { client: { validateUpsertClient: jest.fn() } };

    const res = {
      sendStatus: jest.fn().mockReturnValue({ httpStatus: 201 }),
      error: jest.fn().mockResolvedValue({
        httpStatus: 404,
        ok: true,
        code: 1001,
        errorType: 'Not Found',
        message: 'descripcion del error',
      }),
    };

    const { createClient } = handler({ awsRepositories, getTableName, experts, res });
    await createClient(event);

    expect(experts.client.validateUpsertClient.mock.calls).toMatchSnapshot([
      [
        event,
      ],
    ]);

    expect(awsRepositories.Repository.create.mock.calls).toEqual([
      [
        'devSuggestedProductToUpsert',
        event,
      ],
    ]);

    expect(res.sendStatus.mock.calls).toEqual([]);

    expect(res.error.mock.calls).toEqual([
      [
        'internal_server_error',
        0,
        'server_error',
        500,
      ],
    ]);
  });

});