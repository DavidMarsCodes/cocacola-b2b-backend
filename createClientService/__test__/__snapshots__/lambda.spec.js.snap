// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lambda Function It should answer a validation error in case any of the required parameters are invalid. 1`] = `
Array [
  Array [
    Object {
      "clients": Array [
        Object {
          "address": Object {
            "city": "San Isidro",
            "doorNumber": "227",
            "observations": "El unico lugar que vale la pena",
            "state": "Buenos Aires",
            "street": "Salguero",
            "zipCode": "1609",
          },
          "clientId": "001AR3043-ERP-TEST-A",
          "distributionChannel": "Masivo",
          "erpClientId": "ERP_TEST-A",
          "fantasyName": "TEST-A SA",
          "fiscalId": "227777777722",
          "fiscalName": "Test SA",
          "locks": Array [
            Object {
              "erpLockId": "001AR3043-LOCK-DELETED",
            },
          ],
          "parentClientId": "8888888",
          "salesOrganization": "Test",
        },
      ],
      "countryId": "AR",
      "cpgId": "001",
      "createdTime": Any<String>,
      "organizationId": "3043",
      "transactionId": "111111",
    },
  ],
]
`;

exports[`Lambda Function It should return defualt error 1`] = `
Array [
  Array [
    Object {
      "clients": Array [
        Object {
          "address": Object {
            "city": "San Isidro",
            "doorNumber": "227",
            "observations": "El unico lugar que vale la pena",
            "state": "Buenos Aires",
            "street": "Salguero",
            "zipCode": "1609",
          },
          "clientId": "001AR3043-ERP-TEST-A",
          "distributionChannel": "Masivo",
          "erpClientId": "ERP_TEST-A",
          "fantasyName": "TEST-A SA",
          "fiscalId": "227777777722",
          "fiscalName": "Test SA",
          "locks": Array [
            Object {
              "erpLockId": "001AR3043-LOCK-DELETED",
            },
          ],
          "parentClientId": "8888888",
          "salesOrganization": "Test",
        },
      ],
      "countryId": "AR",
      "cpgId": "001",
      "createdTime": Any<String>,
      "organizationId": "3043",
      "transactionId": "111111",
    },
  ],
]
`;

exports[`Lambda Function You should save an order in the dynamoDB database. 1`] = `
Array [
  Array [
    Object {
      "clients": Array [
        Object {
          "address": Object {
            "city": "San Isidro",
            "doorNumber": "227",
            "observations": "El unico lugar que vale la pena",
            "state": "Buenos Aires",
            "street": "Salguero",
            "zipCode": "1609",
          },
          "clientId": "001AR3043-ERP-TEST-A",
          "distributionChannel": "Masivo",
          "erpClientId": "ERP_TEST-A",
          "fantasyName": "TEST-A SA",
          "fiscalId": "227777777722",
          "fiscalName": "Test SA",
          "locks": Array [
            Object {
              "erpLockId": "001AR3043-LOCK-DELETED",
            },
          ],
          "parentClientId": "8888888",
          "salesOrganization": "Test",
        },
      ],
      "countryId": "AR",
      "cpgId": "001",
      "createdTime": Any<String>,
      "organizationId": "3043",
      "transactionId": "111111",
    },
  ],
]
`;
